File: leetcode/316. Remove Duplicate Letters/main.ts
const removeDuplicateLetters = (s: string): string => {
    const stack: string[]  = [];

    const seen = new Set<string>();
    const lastOcc: Record<string, number> = {};

    for (let i = 0; i < s.length; i++) {
        lastOcc[s[i]] = i
    }

    for (let i = 0; i < s.length; i++) {
        if (!seen.has(s[i])) {
            while (stack.length && s[i] < stack[stack.length-1] && i < lastOcc[stack[stack.length-1]]) {
                seen.delete(stack.pop()!);
            }

            seen.add(s[i])
            stack.push(s[i]);
        }
    }

    return stack.join("");
};

File: leetcode/316. Remove Duplicate Letters/main.go
package main

func removeDuplicateLetters(s string) string {
	stack := []rune{}
	seen := make(map[rune]bool)

	lastOcc := make(map[rune]int)
	for i, c := range s {
		lastOcc[c] = i
	}

	for i, c := range s {
		if !seen[c] {
			for len(stack) > 0 && c < stack[len(stack)-1] && i < lastOcc[stack[len(stack)-1]] {
				delete(seen, stack[len(stack)-1])
				stack = stack[:len(stack)-1]
			}
			seen[c] = true
			stack = append(stack, c)
		}
	}

	return string(stack)
}
